src/components/accounts/SetRecoveryInfo.js

```diff
@@ -0,0 +1,88 @@
+import React, { Component } from 'react'
+import { connect } from 'react-redux'
+import { isValidPhoneNumber } from 'react-phone-number-input'
+
+import { Wallet } from '../../utils/wallet'
+
+import SetRecoveryInfoSection from './SetRecoveryInfoSection'
+import SetRecoveryInfoForm from './SetRecoveryInfoForm'
+import SetRecoveryInfoContainer from './SetRecoveryInfoContainer'
+import { requestCode, validateCode } from '../../actions/account';
+
+class SetRecoveryInfo extends Component {
+   state = {
+      loader: false,
+      phoneNumber: '',
+      isLegit: false,
+   }
+
+   componentDidMount = () => {
+      this.wallet = new Wallet()
+   }
+
+   handleChange = (e, { name, value }) => {
+      this.setState(() => ({
+         [name]: value,
+         isLegit: this.isLegitField(name, value)
+      }))
+   }
+
+   isLegitField(name, value) {
+      // TODO: Use some validation framework?
+      let validators = {
+         phoneNumber: isValidPhoneNumber,
+         securityCode: value => !!value.trim().match(/^\d{6}$/)
+      }
+      return validators[name](value);
+   }
+
+   handleSubmit = e => {
+      e.preventDefault()
+
+      if (!this.state.isLegit) {
+         return false
+      }
+
+      const { dispatch } = this.props;
```

Comment:
Let's start adding tests for the wallet code. This looks like it doesn't have a lot of UI logic, so should be possible to write unit tests for this.

Reply:
I plan to add [integration tests for contract helper](https://github.com/nearprotocol/near-contract-helper/issues/29) and [end-to-end tests for wallet](https://github.com/nearprotocol/near-wallet/issues/33).

Unit tests are going to be easy to write for Redux-based app, but I don't think they are going to be very useful. Most of our real world problems are with broken integrations.